{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4d35a184-789b-4f59-872d-b0fe94f7934e",
   "metadata": {},
   "source": [
    "# Software Information Entropy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "20064744-b294-42d8-8104-a24b8ab12efe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>article_date</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>github_link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>29409532</td>\n",
       "      <td>2018-02-06</td>\n",
       "      <td>SCANPY: large-scale single-cell gene expressio...</td>\n",
       "      <td>Wolf, Angerer, Theis</td>\n",
       "      <td>https://github.com/theislab/Scanpy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29409532</td>\n",
       "      <td>2018-02-06</td>\n",
       "      <td>SCANPY: large-scale single-cell gene expressio...</td>\n",
       "      <td>Wolf, Angerer, Theis</td>\n",
       "      <td>https://github.com/theislab/anndata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30445487</td>\n",
       "      <td>None</td>\n",
       "      <td>G4HunterApps.</td>\n",
       "      <td>Lacroix</td>\n",
       "      <td>https://github.com/LacroixLaurent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>35814290</td>\n",
       "      <td>2022-03-28</td>\n",
       "      <td>WormBase single-cell tools.</td>\n",
       "      <td>da Veiga Beltrame, Arnaboldi, Sternberg</td>\n",
       "      <td>https://github.com/WormBase/scdefg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>35814290</td>\n",
       "      <td>2022-03-28</td>\n",
       "      <td>WormBase single-cell tools.</td>\n",
       "      <td>da Veiga Beltrame, Arnaboldi, Sternberg</td>\n",
       "      <td>https://github.com/WormBase/wormcells-viz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10421</th>\n",
       "      <td>37304664</td>\n",
       "      <td>2023-05-26</td>\n",
       "      <td>Palmprint recognition based on gating mechanis...</td>\n",
       "      <td>Zhang, Xu, Jin, Qi, Yang, Bai</td>\n",
       "      <td>https://github.com/Ywatery/GLnet</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10422</th>\n",
       "      <td>37287133</td>\n",
       "      <td>None</td>\n",
       "      <td>Improving the identification of miRNA-disease ...</td>\n",
       "      <td>He, Qiao, Fang, Bao</td>\n",
       "      <td>https://github.com/qwslle/MTLMDA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10423</th>\n",
       "      <td>37245652</td>\n",
       "      <td>2023-05-26</td>\n",
       "      <td>GAEP: a comprehensive genome assembly evaluati...</td>\n",
       "      <td>Zhang, Lu, Ruan</td>\n",
       "      <td>https://github.com/zy-optimistic/GAEP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10424</th>\n",
       "      <td>37231345</td>\n",
       "      <td>2023-05-25</td>\n",
       "      <td>scMEB: a fast and clustering-independent metho...</td>\n",
       "      <td>Zhu, Yang</td>\n",
       "      <td>https://github.com/FocusPaka/scMEB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10425</th>\n",
       "      <td>37216914</td>\n",
       "      <td>None</td>\n",
       "      <td>MultiNEP: a multi-omics network enhancement fr...</td>\n",
       "      <td>Xu, Marchionni, Wang</td>\n",
       "      <td>https://github.com/Karenxzr/MultiNep</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10426 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           PMID article_date  \\\n",
       "0      29409532   2018-02-06   \n",
       "1      29409532   2018-02-06   \n",
       "2      30445487         None   \n",
       "3      35814290   2022-03-28   \n",
       "4      35814290   2022-03-28   \n",
       "...         ...          ...   \n",
       "10421  37304664   2023-05-26   \n",
       "10422  37287133         None   \n",
       "10423  37245652   2023-05-26   \n",
       "10424  37231345   2023-05-25   \n",
       "10425  37216914         None   \n",
       "\n",
       "                                                   title  \\\n",
       "0      SCANPY: large-scale single-cell gene expressio...   \n",
       "1      SCANPY: large-scale single-cell gene expressio...   \n",
       "2                                          G4HunterApps.   \n",
       "3                            WormBase single-cell tools.   \n",
       "4                            WormBase single-cell tools.   \n",
       "...                                                  ...   \n",
       "10421  Palmprint recognition based on gating mechanis...   \n",
       "10422  Improving the identification of miRNA-disease ...   \n",
       "10423  GAEP: a comprehensive genome assembly evaluati...   \n",
       "10424  scMEB: a fast and clustering-independent metho...   \n",
       "10425  MultiNEP: a multi-omics network enhancement fr...   \n",
       "\n",
       "                                       authors  \\\n",
       "0                         Wolf, Angerer, Theis   \n",
       "1                         Wolf, Angerer, Theis   \n",
       "2                                      Lacroix   \n",
       "3      da Veiga Beltrame, Arnaboldi, Sternberg   \n",
       "4      da Veiga Beltrame, Arnaboldi, Sternberg   \n",
       "...                                        ...   \n",
       "10421            Zhang, Xu, Jin, Qi, Yang, Bai   \n",
       "10422                      He, Qiao, Fang, Bao   \n",
       "10423                          Zhang, Lu, Ruan   \n",
       "10424                                Zhu, Yang   \n",
       "10425                     Xu, Marchionni, Wang   \n",
       "\n",
       "                                     github_link  \n",
       "0             https://github.com/theislab/Scanpy  \n",
       "1            https://github.com/theislab/anndata  \n",
       "2              https://github.com/LacroixLaurent  \n",
       "3             https://github.com/WormBase/scdefg  \n",
       "4      https://github.com/WormBase/wormcells-viz  \n",
       "...                                          ...  \n",
       "10421           https://github.com/Ywatery/GLnet  \n",
       "10422           https://github.com/qwslle/MTLMDA  \n",
       "10423      https://github.com/zy-optimistic/GAEP  \n",
       "10424         https://github.com/FocusPaka/scMEB  \n",
       "10425       https://github.com/Karenxzr/MultiNep  \n",
       "\n",
       "[10426 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# read example data which includes pubmed github links detected from article abstracts\n",
    "pd.read_parquet(\"../../../tests/data/examples/pubmed/pubmed_github_links.parquet\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b60573a4",
   "metadata": {},
   "source": [
    "# Lines of Code Changed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc3cb9f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "import git\n",
    "def calculate_loc_changes(\n",
    "    repo_path: pathlib.Path, source: str, target: str, file_names: list[str]\n",
    ") -> dict[str, int]:\n",
    "    \"\"\"\n",
    "    Finds the total number of code lines changed for each specified file between two commits.\n",
    "\n",
    "    Args:\n",
    "        repo_path (pathlib.Path): The path to the git repository.\n",
    "        source (str): The source commit hash.\n",
    "        target (str): The target commit hash.\n",
    "        file_names (list[str]): List of file names to calculate changes for.\n",
    "\n",
    "    Returns:\n",
    "        dict[str, int]: A dictionary where the key is the filename, and the value is the lines changed (added and removed).\n",
    "    \"\"\"\n",
    "    repo = git.Repo(repo_path)\n",
    "    changes = {}\n",
    "\n",
    "    for file_name in file_names:\n",
    "        # Get the diff output for the file between the two commits\n",
    "        diff_output = repo.git.diff(source, target, \"--numstat\", \"--\", file_name)\n",
    "        # Parse the diff output, then sum the the value of lines added and removed\n",
    "        lines_changed = sum(\n",
    "            abs(int(removed)) + int(added)\n",
    "            for added, removed, _ in (line.split() for line in diff_output.splitlines())\n",
    "        )\n",
    "        changes[file_name] = lines_changed\n",
    "    return changes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f7c6929",
   "metadata": {},
   "source": [
    "# LoC changed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d715ca03",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "986b40f0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
